### ISSUE TITLE
Reduce CPU usage

### ISSUE THREAD

**Issue by Nopik** [vapor-ware/ksync#143]
I started to play with `ksync` today (on Docker For Mac), and noticed, that, it is periodically starting to eat 100% cpu. Not sure what exactly triggers this behavior (I suspect laptop sleep mode) as directly after `ksync watch` start the CPU is not used that much (except of the fact that `ksync watch` start might cause like 30-60 container restarts, which is probably another bug...) - but after a while (4-5 times per day?) laptop fans are working on full speed and the process works on 100% cpu, until I restart it.

**Comment by timfallmk** [comment]
@Nopik Hmm, that's not good, and definitely not what we see here. I'm concerned foremost by the container restarts. The process for `watch` is pretty straightforward, we launch a container in the background with an instance of `syncthing` configured to the scheme you specified in your configuration. The foreground process starts an `grpc` server. If you choose `--daemon`, both processes are launched in the background.

If you're referring to _cluster_ containers (which I guess is also running locally for you), then it gets a bit more tricky. Any change to the filesystem, locally or on the cluster, will cause the target container to restart. The details on this are somewhat annoying, but it's mostly because docker doesn't refresh the mount table outside of a restart and we're using that table to manage directories being synced.

It could be that frequent changes on either side are causing rapid turn over, but it shouldn't be _that_ frequent. How frequently are changes made to either mirrored directory tree? Do you see high CPU usage when not making changes on either side?

As an added wrinkle, the current k8s on docker for mac deployment tends to be a little kludgy and suck up resources, so it's hard to tell what's consuming what.

**Comment by Nopik** [comment]
So, most of the time the dirs on my setup are perfectly synced (I'm not coding 100% of the time, so mostly ksync is just waiting for me to edit some files). Last night I was putting laptop to sleep, when there was nothing to sync again, with `ksync watch` running, on one folder. This morning laptop was warm (clearly was waking up during the night) - and `ksync watch` was at 100% cpu when I unlocked the machine. When I restarted it, following logs were there:

```
INFO[0000] listening                                     bind=127.0.0.1 port=40322
INFO[0000] new pod detected                              pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0003] folder sync running                           pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0006] updating                                      pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0006] updating                                      pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0006] update complete                               pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0006] issuing reload                                pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0007] updating                                      pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0007] update complete                               pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0007] reloaded                                      pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0007] updating                                      pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0007] update complete                               pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0008] updating                                      pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0008] update complete                               pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0008] updating                                      pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0008] update complete                               pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0008] updating                                      pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
INFO[0008] update complete                               pod=pouring-dog-deserializer-75c89b8896-j274b spec=natural-ladybird
```
so, pretty standard, I think. This time it was short, it stopped at `[0008]`. Usually when I do restart `kwatch` it goes to `30` or `58` (highest I've seen) - just issuing reloads one after another on seemingly previously synced dir.

The container is a Node.js app, with its `node_modules` black hole, ~25k files, ~3k dirs, ~134MiB. Yesterday I did added `.stignore` with `node_modules` on it, hoping that it will reduce the problems, but situation didn't change. After ignoring it is like 34 files, 14 dirs, 250KiB.

Answering your questions:
- I was referring to `ksync watch` itself draining CPU, not any containers. I do see it in `top` getting 100% cpu, and ctrl-c + restart fixes the problem for short time (my laptop is often going to sleep when I'm away from keyboard, so maybe the problem happens only on machine sleep)
- I wasn't using any `--daemon` flags, at least not consciously
- the changes are not that frequent. They do come only from me typing in the IDE + `gulp` doing TS to JS conversion. When that happens, usually dir is synced in <2 seconds. I.e. one of the first things I've measured is how fast new code boots up after I press cmd-s in IDE, and the time was pretty good (could be 0.5-1s better, but still not bad). Interestingly, if I just modify some file, without `gulp` doing its compilation phase (which usually takes 2-3 seconds on this particular folder, producing lots of files), then I need to wait like 4-6 seconds until `ksync` notices that there was a change on a filesystem. That alone is pretty annoying, could be nice if I could tweak it somehow.
- typically when there are no filesystem changes, `ksync watch` is not eating cpu. I'm monitoring it in `top` the whole time I'm writing this, it gets like 0.3% cpu or something like that. Until it decides to take 100% and begs for restart, that is.

For what it's worth, for some reason time in my k8s cluster is slightly off, about 56 seconds ahead, not sure why Docker For Mac didn't synced the time properly. And, on top of that it is 1 hour earlier, probably due to different timezone settings.

**Comment by Nopik** [comment]
One detail more, it is not that any laptop sleep is causing the problem. Like, when I do intentionally put laptop to sleep and wake it 2-3 minutes later, `ksync` behaves perfectly.

**Comment by grampelberg** [comment]
Is the time off in your cluster when you intentionally put it to sleep?

**Comment by Nopik** [comment]
I didn't measured it, but I think that the time got synced incorrectly upon k8s install (weeks ago), and got stuck at fixed offset, or at least drifting very slowly.

**Comment by timfallmk** [comment]
@Nopik We think we've tracked down your issue. When I use a local cluster (like Docker for Mac k8s), we can't assume uptime like we did with server side clusters (go figure). The problem arises when the cluster becomes unresponsive (in your case due to a sleep event) and the client half of the sync process, to put it technically, "freaks out". There's debounce and backoff on the entire `watch` process (both sides of it), but apparently not when an already established connect goes dead without a shutdown notice. There may also be a race condition aspect as it only happens when the `watch` process is not _also_ shut down at the same time. I'm guessing we didn't run into this in our testing since we weren't running _long running_ **local** clusters.

We're still tracking down the exact process that is causing the repeated checks, but that should give you some more info about causes in the meantime. Cycling the `watch` process after any such event should clear the problem temporarily.

**Comment by Nopik** [comment]
Great! Thanks for the update. Indeed, that could be the issue. Once I started monitoring it closer, I didn't noticed 100% cpu usage in normal operation, only after (some) laptop sleeps. Ksync is quite nice tool I started to use for my normal workflow (first time really trying to develop on k8s directly), and this is one of the 2 most annoying bits (the second one is that the sync process sometimes takes lot of time to kick in - which maybe is configurable somehow..).

**Comment by timfallmk** [comment]
Good to know! We're happy it brought someone else some use.

If you want, you can file another issue for the sync time. Off the top of my head, I'm not sure why it would take a while, but that could be an internal `syncthing` configuration.

**Comment by Nopik** [comment]
Also, when you'll have a chance to release a version with this bug fixed or at least mitigated?

**Comment by timfallmk** [comment]
@Nopik @grampelberg Is still tracking down why exactly this occurs. We'll try to update you when a fix is merged, then we can cut a new release.

**Comment by Nopik** [comment]
That will be awesome, thanks!

**Comment by timfallmk** [comment]
@Nopik FTFY https://github.com/vapor-ware/ksync/releases/tag/0.1.2
